File: mac_changer.py
#!/usr/bin/env python
import subprocess
import random
import argparse
import re  # Import the regular expression module for MAC address validation

def generate_random_mac():
    return ':'.join(f"{random.randint(10, 99):02x}" for _ in range(6))

def is_valid_mac(mac):
    # Use a regular expression to check if the MAC address is in a valid format
    mac_pattern = re.compile(r'^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})$')
    return bool(mac_pattern.match(mac))

def change_mac(interface, new_mac):
    print(f"[+] Changing MAC address of {interface} to {new_mac}")
    subprocess.call(f"ifconfig {interface} down; ifconfig {interface} hw ether {new_mac}; ifconfig {interface} up", shell=True)

def main():
    parser = argparse.ArgumentParser(description="Change MAC address of a network interface.")
    parser.add_argument("-i", "--interface", help="Specify the network interface (e.g., wlan0, eth0)", required=True)
    parser.add_argument("-m", "--mac", help="Specify the new MAC address")
    parser.add_argument("-r", "--random", help="Use a random MAC address", action="store_true")

    args = parser.parse_args()

    interface = args.interface

    if args.random:
        new_mac = generate_random_mac()
    elif args.mac:
        new_mac = args.mac
        if not is_valid_mac(new_mac):
            print(f"Error: Invalid MAC address format: {new_mac}")
            return
    else:
        new_mac = input("Enter the new MAC address: ")

    change_mac(interface, new_mac)

if __name__ == "__main__":
    main()

File: mac_changer.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>

#define MAC_LENGTH 18

void generate_random_mac(char *random_mac) {
    int i;
    for (i = 0; i < 6; i++) {
        sprintf(random_mac + 3 * i, "%02x:", rand() % 256);
    }
    random_mac[MAC_LENGTH - 1] = '\0'; // Null-terminate the string
}

int is_valid_mac(const char *mac) {
    // MAC address format validation
    int i, len = strlen(mac), count = 0;
    if (len != MAC_LENGTH - 1) return 0; // Check length

    for (i = 0; i < len; i++) {
        if (i % 3 == 2) {
            if (mac[i] != ':') return 0; // Check separator
            count++;
        } else {
            if (!isxdigit(mac[i])) return 0; // Check hex digit
        }
    }
    if (count != 5) return 0; // Check separator count
    return 1;
}

void change_mac(const char *interface, const char *new_mac) {
    char command[100];
    // Construct the command to change the MAC address
    snprintf(command, sizeof(command), "ifconfig %s down && ifconfig %s hw ether %s && ifconfig %s up", interface, interface, new_mac, interface);
    printf("[+] Changing MAC address of %s to %s\n", interface, new_mac);
    system(command); // Execute the command
}

int main(int argc, char *argv[]) {
    char interface[20], new_mac[MAC_LENGTH];
    if (argc < 2) {
        printf("Usage: %s -i <interface> [-m <new_mac> | -r]\n", argv[0]);
        return 1;
    }
    int opt;
    while ((opt = getopt(argc, argv, "i:m:r")) != -1) {
        switch (opt) {
            case 'i':
                strcpy(interface, optarg);
                break;
            case 'm':
                strcpy(new_mac, optarg);
                if (!is_valid_mac(new_mac)) {
                    printf("Error: Invalid MAC address format: %s\n", new_mac);
                    return 1;
                }
                break;
            case 'r':
                generate_random_mac(new_mac);
                break;
            default:
                printf("Usage: %s -i <interface> [-m <new_mac> | -r]\n", argv[0]);
                return 1;
        }
    }
    if (argc == 2) { // Ask for new MAC address if not provided
        printf("Enter the new MAC address: ");
        scanf("%s", new_mac);
        if (!is_valid_mac(new_mac)) {
            printf("Error: Invalid MAC address format: %s\n", new_mac);
            return 1;
        }
    }
    change_mac(interface, new_mac);
    return 0;
}
